src/factory/src/api.rs

```diff
@@ -2,126 +2,135 @@
 //! Copyright  : 2021 InfinitySwap Team
 //! Stability  : Experimental
 
+use std::cell::RefCell;
+use std::rc::Rc;
+
 use crate::error::TokenFactoryError;
 use crate::state::{get_token_bytecode, State};
-use candid::{candid_method, Nat, Principal};
+use candid::{Nat, Principal};
 use common::types::Metadata;
-use ic_cdk_macros::*;
+use ic_canister::{init, query, update, Canister};
 use ic_helpers::factory::error::FactoryError;
 use ic_helpers::factory::FactoryState;
-use ic_storage::IcStorage;
 
-ic_helpers::init_factory_api!(State, crate::state::get_token_bytecode());
+mod inspect_message;
 
-#[init]
-#[candid_method(init)]
-pub fn init(controller: Principal, ledger_principal: Option<Principal>) {
-    State::new(controller, ledger_principal).reset();
-}
+ic_helpers::extend_with_factory_api!(
+    TokenFactoryCanister,
+    state,
+    crate::state::get_token_bytecode()
+);
 
-#[inspect_message]
-fn inspect_message_function() {
-    if ic_cdk::api::call::method_name() == "set_token_bytecode" {
-        return ic_cdk::api::call::accept_message();
-    }
+#[derive(Clone, Canister)]
+pub struct TokenFactoryCanister {
+    #[id]
+    principal: Principal,
 
-    match &State::get().borrow().token_wasm {
-        Some(_) => ic_cdk::api::call::accept_message(),
-        None => ic_cdk::api::call::reject("the factory hasn't been completely intialized yet"),
-    }
+    #[state]
+    state: Rc<RefCell<State>>,
 }
 
-/// Returns the token, or None if it does not exist.
-#[query(name = "get_token")]
-#[candid_method(query, rename = "get_token")]
-async fn get_token(name: String) -> Option<Principal> {
-    State::get().borrow().factory.get(&name)
-}
-
-pub async fn set_token_bytecode_impl(bytecode: Vec<u8>) {
-    State::get().borrow_mut().token_wasm.replace(bytecode);
-}
-
-#[update(name = "set_token_bytecode")]
-#[candid_method(update, rename = "set_token_bytecode")]
-pub async fn set_token_bytecode(bytecode: Vec<u8>) {
-    if State::get().borrow().controller() != ic_cdk::api::caller() {
-        ic_cdk::api::call::reject("not authorized");
-        return;
+#[allow(dead_code)]
```

Comment:
I don't think that we want to decorate everything with this attribute. What part of the implementation requires this and why? Can we deal with it in a better way?

Reply:
I'm looking at this now, it's not immediately clear why this is generating warnings